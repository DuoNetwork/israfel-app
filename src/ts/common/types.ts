import { AnyAction } from 'redux';
import { ThunkAction } from 'redux-thunk';
export * from '../../../../israfel-relayer/src/common/types';
import * as relayerTypes from '../../../../israfel-relayer/src/common/types';

export type VoidThunkAction = ThunkAction<void, IState, undefined, AnyAction>;

export interface IState {
	readonly dex: IDexState;
	readonly dynamo: IDynamoState;
	readonly ui: IUIState;
	readonly web3: IWeb3State;
	readonly ws: IWsState;
}

export interface IDexState {
	readonly userOrders: relayerTypes.IUserOrder[];
	readonly orderBookSnapshot: relayerTypes.IOrderBookSnapshot;
	readonly orderBookSubscription: string;
	readonly userSubscription: number;
	readonly tokenBalance: ITokenBalance;
}

export interface IDynamoState {
	readonly status: relayerTypes.IStatus[];
}

export interface IUIState {
	readonly locale: string;
}

export interface IWeb3State {
	readonly account: string;
	readonly network: number;
	readonly ethBalance: IEthBalance;
}

export interface IWsState {
	readonly connection: boolean;
	readonly tokens: relayerTypes.IToken[];
}

export interface IEthBalance {
	eth: number;
	weth: number;
	allowance: number;
}

export interface ITokenBalance {
	balance: number;
	allowance: number;
}
