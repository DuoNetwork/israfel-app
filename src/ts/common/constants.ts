export * from '../../../../israfel-relayer/src/common/constants';

export const AC_ACCOUNT = 'account';
export const AC_NETWORK = 'network';
// export const AC_GAS_PX = 'gasPrice';
export const AC_INFO = 'info';
export const AC_LOCALE = 'locale';
export const AC_ORDER = 'order';
export const AC_ORDER_HISTORY = 'orderHistory';
export const AC_CONNECTION = 'connection';
export const AC_OB_UPDATE = 'orderBookUpdate';
export const AC_OB_SNAPSHOT = 'orderBookSnapshot';
export const AC_OB_SUB = 'orderBookSubscription';
export const AC_USER_SUB = 'userSubscription';
export const AC_ETH_BALANCE = 'ethBalance';
export const AC_TOKEN_BALANCE = 'tokenBalance';

export const LOCALE_CN = 'CN';
export const LOCALE_EN = 'EN';
export const LOCALE_RU = 'RU';
export const LOCALE_JP = 'JP';
export interface ILocaleText {
	EN: string;
	CN: string;
	RU: string;
	JP: string;
	[key: string]: string;
}
export const TH_ETH = 'Eth';
export const TH_WETH = 'Weth';
export const TH_ALLOWANCE = 'Allowance';
export const TH_SUBMIT = 'Submit';
export const TH_APPROVE = 'Approve';
export const TH_CLEAR = 'Clear';
export const TH_ORDER = 'Order';
export const TH_HISTORY = 'History';
export const TH_LOADING = 'Loading';
export const TH_TRADE = 'Trades';
export const TH_SELL = 'Sell';
export const TH_BUY = 'Buy';
export const TH_PX = 'Price';
export const TH_AMT = 'Amount';
export const TH_BALANCE = 'Balance';
export const TH_FILL = 'Fill';
export const TH_SIDE = 'Side';
export const TH_FEE = 'Fee';
export const TH_STATUS = 'Status';
export const TH_ACTIONS = 'Actions';
export const TH_EXPIRY = 'Expiry';
export const TH_ORDER_HASH = 'Order Hash';
export const TH_TIME = 'Time';
export const TH_WRAP = 'Wrap';
export const TH_UNWRAP = 'Unwrap';
export const TH_PROCESS = 'Process';
export const TH_UPDATED = 'Updated';
export const TH_INFO = 'Info';
export const TH_HOSTNAME = 'Hostname';
export const TH_COLLATERAL = 'Collateral';
export const TH_CONVERT = 'Convert';
export const TH_CREATE = 'Create';
export const TH_REDEEM = 'Redeem';
export const TH_PLACEORDER = 'Place Order';

export const TT_DELETE_ORDER = 'Are you sure to delete this order?';
export const TT_NETWORK_CHECK: ILocaleText = {
	[LOCALE_CN]:
		'此页面只支持' + (__KOVAN__ ? 'KOVAN' : 'MainNet') + '，请在MetaMask中选择正确的网络',
	[LOCALE_EN]:
		'This page is built for ' +
		(__KOVAN__ ? 'KOVAN' : 'MainNet') +
		', please choose the correct network in MetaMask',
	[LOCALE_JP]:
		'このページは' +
		(__KOVAN__ ? 'KOVAN' : 'MainNet') +
		'のために作られています。メタマスクに、正しいネットワークを選んでください',
	[LOCALE_RU]:
		'Данная страница предназначена для ' +
		(__KOVAN__ ? 'KOVAN' : 'MainNet') +
		', выберите необходимую сеть в MetaMask'
};
