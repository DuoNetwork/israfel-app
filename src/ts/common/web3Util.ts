import { Constants, Web3Util } from '../../../../israfel-common/src';

const web3Util = new Web3Util(window, __ENV__ === Constants.DB_LIVE, '', false);
export default web3Util;
