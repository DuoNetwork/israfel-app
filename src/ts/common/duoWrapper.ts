// fix for @ledgerhq/hw-transport-u2f 4.28.0
import '@babel/polyfill';
import DualClassWrapper from '../../../../duo-contract-wrapper/src/DualClassWrapper';
import Web3Wrapper from '../../../../duo-contract-wrapper/src/Web3Wrapper';
import * as CST from './constants';

let infura = {
	token: ''
};
try {
	infura = require('../../../../israfel-relayer/src/keys/infura.json');
} catch (error) {
	console.log(error);
}
const provider =
	(__ENV__ === CST.DB_LIVE ? CST.PROVIDER_INFURA_MAIN : CST.PROVIDER_INFURA_KOVAN) +
	'/' +
	infura.token;
export const duoWeb3Wrapper = new Web3Wrapper(window, '', provider, __ENV__ === CST.DB_LIVE);
export const dualClassWrappers: { [custodian: string]: DualClassWrapper } = {};
export const getDualClassWrapper = (custodian: string) => {
	if (!dualClassWrappers[custodian])
		dualClassWrappers[custodian] = new DualClassWrapper(duoWeb3Wrapper, custodian);
	return dualClassWrappers[custodian];
};
