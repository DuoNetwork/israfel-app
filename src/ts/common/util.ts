import { Constants, IUserOrder, Util as CommonUtil } from '@finbook/israfel-common';
import * as d3 from 'd3';
import moment from 'moment';
import * as CST from './constants';
import { ICustodianInfo } from './types';

class Util {
	public convertUpdateTime(timestamp: number): string {
		const diff = CommonUtil.getUTCNowTimestamp() - timestamp;
		if (diff < 60000) return 'just now';
		else if (diff < 3600000) return Math.floor(diff / 60000) + ' min ago';
		else if (diff < 86400000) return Math.floor(diff / 3600000) + ' hrs ago';
		else return 'long time ago';
	}

	public formatExpiry(isWeek: boolean) {
		return moment(CommonUtil.getExpiryTimestamp(isWeek)).format('YYYY-MM-DD HH:mm');
	}

	public formatTime(time: any) {
		return moment(time).format('MM-DD HH:mm:ss');
	}

	public formatTimeSecond(time: number) {
		return moment(time).format('YYYY-MM-DD HH:mm:ss');
	}

	public formatPriceShort(num: number) {
		if (num < 1) return d3.format('.3f')(num);
		return d3
			.format('.4s')(num)
			.toUpperCase()
			.replace(/G/g, 'B');
	}

	public formatStrike(num: number) {
		return d3.format('.4f')(num);
	}

	public formatBalance(num: number) {
		if (Math.abs(num) < 1e-4) return '0.000';
		return d3
			.format(Math.abs(num) > 1 ? ',.4s' : ',.4n')(num)
			.toUpperCase()
			.replace(/G/g, 'B');
	}

	public formatPercent(num: number) {
		return d3.format('.0%')(num);
	}

	public formatPercentAcc(num: number) {
		return d3.format('.2%')(num);
	}

	public formatNumber(num: number) {
		if (Math.abs(num) < 1e-4) return '0.000';
		if (Math.abs(num) < 1) return d3.format(',.4n')(num);
		if (Math.abs(num) < 100000) return d3.format(',.2f')(num);
		return d3
			.format(',.4s')(num)
			.toUpperCase()
			.replace(/G/g, 'B');
	}

	public formatMaturity(maturity: number) {
		return maturity ? moment(maturity).format('YYYY-MM-DD HH:mm') : CST.TH_PERPETUAL;
	}

	public getOrderTitle(order: IUserOrder) {
		if (order.type === Constants.DB_ADD) return 'Order Placed';
		else if (order.type === Constants.DB_TERMINATE)
			if (order.status === Constants.DB_FILL) return 'Order Fully Filled';
			else if (order.status === Constants.DB_TERMINATE) return 'Order Expired';
			// order.status === Constants.DB_CONFIRMED
			else return 'Order Cancelled';
		else if (order.status === Constants.DB_MATCHING) return 'Order is Being Filled';
		// if (order.status === Constants.DB_PFILL)
		else return 'Order Partially Filled';
	}

	public getOrderFullDescription(order: IUserOrder) {
		const [code1, code2] = order.pair.split('|');
		let baseDescription = `${order.side === Constants.DB_BID ? CST.TH_BUY : CST.TH_SELL} ${
			order.amount
		} ${code1} at ${order.price} ${code2} per ${code1}.`;
		if (order.type === Constants.DB_ADD)
			return baseDescription + ` Valid till ${this.formatMaturity(order.expiry)}`;

		if (order.type === Constants.DB_TERMINATE && order.status === Constants.DB_FILL)
			return baseDescription + ' Fully filled';

		if (order.fill || order.matching) {
			const totalFill =
				order.fill + (order.type === Constants.DB_TERMINATE ? 0 : order.matching);
			baseDescription += ` ${totalFill}(${this.formatPercent(
				totalFill / order.amount
			)}) filled.`;
		}

		if (order.matching && order.type !== Constants.DB_TERMINATE)
			baseDescription += 'Pending settlement';

		if (order.type === Constants.DB_TERMINATE && order.status === Constants.DB_CONFIRMED)
			return baseDescription + ' Cancelled by user.';

		if (order.type === Constants.DB_TERMINATE && order.status === Constants.DB_TERMINATE)
			return baseDescription + ' Expired.';

		if (order.type === Constants.DB_TERMINATE && order.status === Constants.DB_BALANCE)
			return baseDescription + ` Cancelled due to insufficient balance.`;

		if (order.type === Constants.DB_TERMINATE && order.status === Constants.DB_MATCHING)
			return baseDescription + ` Cancelled due to settlement error.`;

		if (order.type === Constants.DB_TERMINATE && order.status === Constants.DB_RESET)
			return baseDescription + ` Cancelled due to custodian reset.`;

		return baseDescription;
	}

	public getOrderDescription(order: IUserOrder) {
		const [code1, code2] = order.pair.split('|');
		return `${order.side === Constants.DB_BID ? CST.TH_BUY : CST.TH_SELL} ${
			order.amount
		} ${code1} at ${order.price} ${code2} per ${code1}. Total ${util.formatBalance(
			order.fill + (order.type === Constants.DB_TERMINATE ? 0 : order.matching)
		)} filled.`;
	}

	public getVersionDescription(order: IUserOrder) {
		if (order.type === Constants.DB_ADD) return 'Order placed.';

		if (order.type === Constants.DB_UPDATE) {
			let description = `Total ${this.formatBalance(order.fill + order.matching)} filled`;
			if (order.matching) description += ', pending settlement';
			return description;
		}

		if (order.type === Constants.DB_TERMINATE)
			if (order.status === Constants.DB_TERMINATE) return 'Expired';
			else if (order.status === Constants.DB_BALANCE)
				return 'Cancelled due to insufficent balance';
			else if (order.status === Constants.DB_MATCHING)
				return 'Cancelled due to settlement error';
			else if (order.status === Constants.DB_RESET) return 'Cancelled due to custodian reset';
			else if (order.status === Constants.DB_FILL) return 'Fully filled';
			// (order.status === Constants.DB_CONFIRMED)
			else return 'Cancelled by user';
		return 'Invalid order';
	}

	public getMaturityDescription(info?: ICustodianInfo) {
		return `This contract ${
			info && info.code.endsWith('PPT')
				? 'never expires'
				: 'expires in ' +
				(info
					? Math.floor(
							(info.states.maturity - CommonUtil.getUTCNowTimestamp()) / 86400000
						).toFixed(0)
					: 0) +
				' days'
		}.`;
	}

	public convertOrdersToCSV(orders: { [pair: string]: IUserOrder[] }): string {
		const headers = [
			CST.TH_PAIR,
			CST.TH_ORDER_HASH,
			CST.TH_VERSION,
			CST.TH_TIME_UTC,
			CST.TH_SIDE,
			CST.TH_PRICE,
			CST.TH_AMOUNT,
			CST.TH_FILL,
			CST.TH_FEE,
			CST.TH_EXPIRY_UTC,
			CST.TH_STATUS,
			CST.TH_TX_HASH
		];
		let output = '';
		headers.forEach(
			(header, index) =>
				(output += this.wrapCSVString(header) + (index < headers.length - 1 ? ',' : '\n'))
		);
		const ordersByHash: { [orderHash: string]: IUserOrder[] } = {};
		for (const pair in orders)
			orders[pair].forEach(order => {
				if (!ordersByHash[order.orderHash]) ordersByHash[order.orderHash] = [];
				ordersByHash[order.orderHash].push(order);
			});

		const rows: any[][] = [];
		for (const orderHash in ordersByHash) {
			ordersByHash[orderHash].sort((a, b) => a.currentSequence - b.currentSequence);
			ordersByHash[orderHash].forEach((order, index) => {
				const row: any[] = [];
				row.push(order.createdAt);
				row.push(order.pair.replace('|', '-')); // CST.TH_PAIR,
				row.push(order.orderHash); // CST.TH_ID,
				row.push(index + 1); // CST.TH_VERSION,
				row.push(
					moment.utc(order.updatedAt || order.createdAt).format('YYYY-MM-DD HH:mm:ss')
				); // CST.TH_TIME_UTC,
				row.push(order.side === Constants.DB_BID ? CST.TH_BUY : CST.TH_SELL); // CST.TH_SIDE,
				row.push(order.price); // CST.TH_PRICE,
				row.push(order.amount); // CST.TH_AMOUNT,
				row.push(order.fill + order.matching); // CST.TH_FILL,
				row.push(order.fee + ' ' + order.feeAsset); // CST.TH_FEE,
				row.push(moment.utc(order.expiry).format('YYYY-MM-DD HH:mm:ss')); // CST.TH_EXPIRY,
				// CST.TH_STATUS,
				if (order.type === Constants.DB_TERMINATE)
					if (order.status === Constants.DB_BALANCE)
						row.push('Cancelled due to insufficent balance');
					else if (order.status === Constants.DB_MATCHING)
						row.push('Cancelled due to settlement error');
					else if (order.status === Constants.DB_RESET)
						row.push('Cancelled due to custodian reset');
					else if (order.status === Constants.DB_FILL) row.push('Fully filled');
					else if (order.status === Constants.DB_CONFIRMED) row.push('Cancelled by user');
					else row.push('Expired');
				else if (order.status === Constants.DB_MATCHING) row.push('Pending settlement');
				else row.push('Open');
				row.push(order.transactionHash || ''); // CST.TH_TX_HASH
				rows.push(row);
			});
		}

		rows.sort((a, b) => a[1].localeCompare(b[1]) || -a[0] + b[0] || -a[3] + b[3]);
		rows.forEach(row =>
			row.forEach((item, index) => {
				if (index)
					output +=
						this.wrapCSVString(item + '') + (index !== row.length - 1 ? ',' : '\n');
			})
		);

		return output;
	}

	public wrapCSVString(input: string): string {
		return ('' + input).indexOf(',') >= 0 ? '"' + input + '"' : input;
	}

	private getEtherScanLink() {
		return `https://${__ENV__ === Constants.DB_LIVE ? '' : 'kovan.'}etherscan.io`;
	}

	public getEtherScanTransactionLink(txHash: string) {
		return `${this.getEtherScanLink()}/tx/${txHash}`;
	}

	public getEtherScanTokenLink(tokenAddr: string, account: string = '') {
		return `${this.getEtherScanLink()}/token/${tokenAddr}${account ? `?a=${account}` : ''}`;
	}

	public getEtherScanAddressLink(account: string) {
		return `${this.getEtherScanLink()}/address/${account}`;
	}

	public fillZero(num: number) {
		return num < 10 ? `0${num}` : `${num}`;
	}

	public isEmptyObject(obj: object | undefined | null): boolean {
		if (!obj) return true;

		for (const prop in obj) if (obj.hasOwnProperty(prop)) return false;

		return true;
	}
}

const util = new Util();
export default util;
