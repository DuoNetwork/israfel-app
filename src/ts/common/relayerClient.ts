import RelayerClient from '../../../../israfel-relayer/src/client/RelayerClient';
import web3Util from './web3Util';

const relayerClient = new RelayerClient(web3Util, __ENV__);
export default relayerClient;
