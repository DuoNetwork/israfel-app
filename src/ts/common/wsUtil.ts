import WebSocket from 'isomorphic-ws';
import * as CST from '../../../../israfel-relayer/src/common/constants';
import {
	IWsOrderResponse,
	IWsUserOrderResponse
} from '../../../../israfel-relayer/src/common/types';
import Web3Util from '../../../../israfel-relayer/src/utils/Web3Util';
import { IWsAddOrderRequest } from '../common/types';
import util from './util';

class WsUtil {
	public ws: WebSocket | null = null;
	private web3Util: Web3Util | null = null;
	private handleOrderBooksUpdate: ((orderBooks: IWsUserOrderResponse) => any) | null = null;
	public init(host: string) {
		this.web3Util = new Web3Util(window, !__DEV__);
		this.ws = new WebSocket(host);
		this.ws.onopen = function open() {
			console.log('Connected');
		};
		this.ws.onmessage = (m: any) => this.handleMessage(m.data.toString());
	}

	public handleMessage(message: string) {
		const wsMsg: IWsOrderResponse = JSON.parse(message);
		const method = wsMsg.method;
		switch (method) {
			case CST.DB_ADD:
				if (this.handleOrderBooksUpdate) {
					const obRes = wsMsg as IWsUserOrderResponse;
					// const src = others[0];
					if (obRes && obRes.userOrder) {
						// obRes.delay = util.getUTCNowTimestamp() - obRes.timestamp;
						console.log('obRes');
						this.handleOrderBooksUpdate(obRes);
					}
				}
				break;
			default:
				break;
		}
	}

	public subscribe(name: string, marketId: string) {
		if (this.ws && name && marketId) {
			const req = {
				type: 'subscribe',
				channel: {
					name: name,
					marketId: marketId
				},
				requestId: Date.now().toString()
			};
			this.ws.send(JSON.stringify(req));
		}
	}

	public async addOrder(zrxAmt: number, ethAmt: number, isBid: boolean) {
		if (!this.web3Util) throw new Error('error');
		const zrxTokenAddress = this.web3Util.getTokenAddressFromName(CST.TOKEN_ZRX);
		const etherTokenAddress = this.web3Util.getTokenAddressFromName(CST.TOKEN_WETH);
		if (etherTokenAddress === undefined) throw console.error('undefined etherTokenAddress');

		const accounts = await this.web3Util.web3Wrapper.getAvailableAddressesAsync();
		const rawOrder = await this.web3Util.createRawOrder(
			accounts[0],
			__DEV__ ? CST.RELAYER_ADDR_KOVAN : CST.RELAYER_ADDR_MAIN,
			isBid ? zrxTokenAddress : etherTokenAddress,
			isBid ? etherTokenAddress : zrxTokenAddress,
			isBid ? zrxAmt : ethAmt,
			isBid ? ethAmt : zrxAmt,
			Math.ceil(util.getUTCNowTimestamp() / 1000) + 3600
		);
		const pair = 'ZRX-WETH';
		const msg: IWsAddOrderRequest = {
			method: CST.DB_ADD,
			channel: CST.DB_ORDERS,
			pair: pair,
			orderHash: rawOrder.orderHash,
			order: rawOrder.signedOrder
		};
		this.ws.send(JSON.stringify(msg));
	}

	public onOrderBooks(handleOrderBooksUpdate: (orderBooks: IWsUserOrderResponse) => any) {
		this.handleOrderBooksUpdate = handleOrderBooksUpdate;
	}
}

const wsUtil = new WsUtil();
export default wsUtil;
