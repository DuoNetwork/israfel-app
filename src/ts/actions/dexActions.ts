import * as CST from 'ts/common/constants';
import {
	IOrderBookSnapshot,
	IOrderBookSnapshotUpdate,
	ITokenBalance,
	IUserOrder,
	VoidThunkAction
} from 'ts/common/types';
import util from 'ts/common/util';
import web3Util from 'ts/common/web3Util';
import wsUtil from 'ts/common/wsUtil';
import dynamoUtil from '../../../../israfel-relayer/src/utils/dynamoUtil';

export function userOrderUpdate(userOrder: IUserOrder) {
	return {
		type: CST.AC_USER_ORDER,
		value: userOrder
	};
}

export function userOrderListUpdate(userOrders: IUserOrder[]) {
	return {
		type: CST.AC_USER_ORDER_LIST,
		value: userOrders
	};
}

export function orderBookUpdate(obUpdate: IOrderBookSnapshotUpdate) {
	return {
		type: CST.AC_OB_UPDATE,
		value: obUpdate
	};
}

export function orderBookSubscriptionUpdate(pair: string) {
	return {
		type: CST.AC_OB_SUB,
		value: pair
	};
}

export function getUserOrders(pair: string): VoidThunkAction {
	return async (dispatch, getState) => {
		const account = getState().web3.account;
		if (account !== CST.DUMMY_ADDR) {
			const now = util.getUTCNowTimestamp();
			dispatch(
				userOrderListUpdate(
					await dynamoUtil.getUserOrders(account, now - 30 * 86400000, now, pair)
				)
			);
		}
	};
}

export function tokenBalanceUpdate(balance: ITokenBalance) {
	return {
		type: CST.AC_TOKEN_BALANCE,
		value: balance
	};
}

export function getTokenBalance(pair: string): VoidThunkAction {
	return async (dispatch, getState) => {
		const account = getState().web3.account;
		const code = pair.split('|')[0];
		if (account !== CST.DUMMY_ADDR)
			dispatch(
				tokenBalanceUpdate({
					balance: await web3Util.getTokenBalance(code, account),
					allowance: await web3Util.getProxyTokenAllowance(code, account)
				})
			);
	};
}

export function orderBookSnapshotUpdate(orderBook: IOrderBookSnapshot) {
	return {
		type: CST.AC_OB_SNAPSHOT,
		value: orderBook
	};
}

export function userSubscriptionUpdate(intervalId: number) {
	return {
		type: CST.AC_USER_SUB,
		value: intervalId
	};
}

export function subscribe(pair: string): VoidThunkAction {
	return dispatch => {
		dispatch(orderBookSubscriptionUpdate(''));
		dispatch(userSubscriptionUpdate(0));
		dispatch(getUserOrders(pair));
		dispatch(getTokenBalance(pair));
		dispatch(orderBookSubscriptionUpdate(pair));
		wsUtil.subscribeOrderBook(pair);
		dispatch(
			userSubscriptionUpdate(
				window.setInterval(() => {
					dispatch(getTokenBalance(pair));
					dispatch(getUserOrders(pair));
				}, 30000)
			)
		);
	};
}
