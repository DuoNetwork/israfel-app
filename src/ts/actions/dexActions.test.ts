import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import * as CST from 'ts/common/constants';
import util from 'ts/common/util';
import web3Util from 'ts/common/web3Util';
import wsUtil from 'ts/common/wsUtil';
import dynamoUtil from '../../../../israfel-relayer/src/utils/dynamoUtil';
import * as dexActions from './dexActions';

const mockStore = configureMockStore([thunk]);

describe('actions', () => {
	test('userOrder', () => {
		expect(dexActions.userOrderUpdate({ test: 'test' } as any)).toMatchSnapshot();
	});

	test('userOrderListUpdate', () => {
		expect(dexActions.userOrderListUpdate([{ test: 'test' }] as any)).toMatchSnapshot();
	});

	test('userSubscriptionUpdate', () => {
		expect(dexActions.userSubscriptionUpdate(123)).toMatchSnapshot();
	});

	test('orderBookSnapshotUpdate', () => {
		expect(dexActions.orderBookSnapshotUpdate({ test: 'test' } as any)).toMatchSnapshot();
	});

	test('orderBookUpdate', () => {
		expect(dexActions.orderBookUpdate({ test: 'test' } as any)).toMatchSnapshot();
	});

	test('orderBookSubscriptionUpdate', () => {
		expect(dexActions.orderBookSubscriptionUpdate('pair')).toMatchSnapshot();
	});

	test('getUserOrders dummy addr', () => {
		const store: any = mockStore({
			web3: {
				account: CST.DUMMY_ADDR
			}
		});
		dynamoUtil.getUserOrders = jest.fn(() =>
			Promise.resolve([
				{
					test: 'test'
				}
			])
		);
		store.dispatch(dexActions.getUserOrders('pair'));
		return new Promise(resolve =>
			setTimeout(() => {
				expect(store.getActions()).toMatchSnapshot();
				expect(dynamoUtil.getUserOrders as jest.Mock).not.toBeCalled();
				resolve();
			}, 0)
		);
	});

	test('getUserOrders', () => {
		util.getUTCNowTimestamp = jest.fn(() => 1234567890);
		const store: any = mockStore({
			web3: {
				account: '0xAccount'
			}
		});
		dynamoUtil.getUserOrders = jest.fn(() =>
			Promise.resolve([
				{
					test: 'test'
				}
			])
		);
		store.dispatch(dexActions.getUserOrders('pair'));
		return new Promise(resolve =>
			setTimeout(() => {
				expect(store.getActions()).toMatchSnapshot();
				expect((dynamoUtil.getUserOrders as jest.Mock).mock.calls).toMatchSnapshot();
				resolve();
			}, 0)
		);
	});

	test('tokenBalanceUpdate', () => {
		expect(
			dexActions.tokenBalanceUpdate({
				balance: 123,
				allowance: 456
			})
		).toMatchSnapshot();
	});

	test('getTokenBalance dummy addr', () => {
		const store: any = mockStore({
			web3: {
				account: CST.DUMMY_ADDR
			}
		});
		web3Util.getTokenBalance = jest.fn(() => Promise.resolve(111));
		web3Util.getProxyTokenAllowance = jest.fn(() => Promise.resolve(222));
		store.dispatch(dexActions.getTokenBalance('code1|code2'));
		return new Promise(resolve =>
			setTimeout(() => {
				expect(store.getActions()).toMatchSnapshot();
				expect(web3Util.getTokenBalance as jest.Mock).not.toBeCalled();
				expect(web3Util.getProxyTokenAllowance as jest.Mock).not.toBeCalled();
				resolve();
			}, 0)
		);
	});

	test('getTokenBalance', () => {
		const store: any = mockStore({
			web3: {
				account: '0xAccount'
			}
		});
		web3Util.getTokenBalance = jest.fn(() => Promise.resolve(111));
		web3Util.getProxyTokenAllowance = jest.fn(() => Promise.resolve(222));
		store.dispatch(dexActions.getTokenBalance('code1|code2'));
		return new Promise(resolve =>
			setTimeout(() => {
				expect(store.getActions()).toMatchSnapshot();
				expect((web3Util.getTokenBalance as jest.Mock).mock.calls).toMatchSnapshot();
				expect((web3Util.getProxyTokenAllowance as jest.Mock).mock.calls).toMatchSnapshot();
				resolve();
			}, 0)
		);
	});

	test('subscribe', () => {
		window.setInterval = jest.fn(() => 123);
		util.getUTCNowTimestamp = jest.fn(() => 1234567890);
		const store: any = mockStore({
			web3: {
				account: '0xAccount'
			}
		});
		dynamoUtil.getUserOrders = jest.fn(() =>
			Promise.resolve([
				{
					test: 'test'
				}
			])
		);
		wsUtil.subscribeOrderBook = jest.fn();
		store.dispatch(dexActions.subscribe('pair'));
		return new Promise(resolve =>
			setTimeout(() => {
				expect(store.getActions()).toMatchSnapshot();
				expect((wsUtil.subscribeOrderBook as jest.Mock).mock.calls).toMatchSnapshot();
				expect((dynamoUtil.getUserOrders as jest.Mock).mock.calls).toMatchSnapshot();
				resolve();
			}, 0)
		);
	});
});
