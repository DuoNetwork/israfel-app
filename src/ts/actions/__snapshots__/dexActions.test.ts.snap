// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions getTokenBalance 1`] = `
Array [
  Object {
    "type": "tokenBalance",
    "value": Object {
      "allowance": 222,
      "balance": 111,
    },
  },
]
`;

exports[`actions getTokenBalance 2`] = `
Array [
  Array [
    "code1",
    "0xAccount",
  ],
]
`;

exports[`actions getTokenBalance 3`] = `
Array [
  Array [
    "code1",
    "0xAccount",
  ],
]
`;

exports[`actions getTokenBalance dummy addr 1`] = `Array []`;

exports[`actions getUserOrders 1`] = `
Array [
  Object {
    "type": "userOrderList",
    "value": Array [
      Object {
        "test": "test",
      },
    ],
  },
]
`;

exports[`actions getUserOrders 2`] = `
Array [
  Array [
    "0xAccount",
    -1357432110,
    1234567890,
    "pair",
  ],
]
`;

exports[`actions getUserOrders dummy addr 1`] = `Array []`;

exports[`actions orderBookSnapshotUpdate 1`] = `
Object {
  "type": "orderBookSnapshot",
  "value": Object {
    "test": "test",
  },
}
`;

exports[`actions orderBookSubscriptionUpdate 1`] = `
Object {
  "type": "orderBookSubscription",
  "value": "pair",
}
`;

exports[`actions orderBookUpdate 1`] = `
Object {
  "type": "orderBookUpdate",
  "value": Object {
    "test": "test",
  },
}
`;

exports[`actions subscribe 1`] = `
Array [
  Object {
    "type": "orderBookSubscription",
    "value": "",
  },
  Object {
    "type": "userSubscription",
    "value": 0,
  },
  Object {
    "type": "orderBookSubscription",
    "value": "pair",
  },
  Object {
    "type": "userSubscription",
    "value": 123,
  },
  Object {
    "type": "userOrderList",
    "value": Array [
      Object {
        "test": "test",
      },
    ],
  },
  Object {
    "type": "tokenBalance",
    "value": Object {
      "allowance": 222,
      "balance": 111,
    },
  },
]
`;

exports[`actions subscribe 2`] = `
Array [
  Array [
    "pair",
  ],
]
`;

exports[`actions subscribe 3`] = `
Array [
  Array [
    "0xAccount",
    -1357432110,
    1234567890,
    "pair",
  ],
]
`;

exports[`actions tokenBalanceUpdate 1`] = `
Object {
  "type": "tokenBalance",
  "value": Object {
    "allowance": 456,
    "balance": 123,
  },
}
`;

exports[`actions userOrder 1`] = `
Object {
  "type": "userOrder",
  "value": Object {
    "test": "test",
  },
}
`;

exports[`actions userOrderListUpdate 1`] = `
Object {
  "type": "userOrderList",
  "value": Array [
    Object {
      "test": "test",
    },
  ],
}
`;

exports[`actions userSubscriptionUpdate 1`] = `
Object {
  "type": "userSubscription",
  "value": 123,
}
`;
